### PR TITLE
Component Model + PoC Mob

### PR DESCRIPTION
**Pull request by N3X15**
# Executive Summary
This PR adds a framework for a generalized software component system.  

# Details
Many of us are familiar with the concept of Object-Oriented Programming (OOP).  In short, it breaks a task up into interactions between entities (objects) and their relations.  Space Station 13 already makes have use of this concept, but there's one problem:  Many of the entities in SS13 can be enormous and complex, with human entities having a Life() process that encompasses a massive file of its own, and requires grey-beard levels of wizardry to work on it successfully.

There's a way to simplify OOP further: by using Components.

This PR adds a basic component model, which consists of a Container (/datum/component_container), which can hold one or more Components (/datum/component).  This allows us to break a big, monolithic and complex process like Life() into smaller, more manageable chunks.  More importantly, components can communicate with each other without direct references, allowing for highly flexible and adaptable systems.

As a demonstration, I have created a giant component spider mob which uses components for AI.  These components communicate with a mob controller, but can also communicate with an object controller, making these components more reusable and adaptable than our existing mob framework.

### CODE REVIEW

**Comment by PJB3005** [comment]
Fucking.

Finally.

File: code/__HELPERS/maths.dm [hunk 1]
```diff
@@ -226,6 +226,9 @@ proc/arctan(x)
 #undef k1
 #undef k2
 
+/proc/get_dist_squared(var/atom/a, var/atom/b)
```
Comment (PJB3005) [inline code/__HELPERS/maths.dm]: The amount of pythag functions in the code by now must be tremendous.

File: code/modules/components/ai/ai_component.dm [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+	var/state=0 // AI_STATE_* of the AI.
+
+/datum/component/ai/RecieveSignal(var/message_type, var/list/args)
+	switch(message_type)
+		if("state") // list("name"="statename")
+			state = args["name"]
+
+/datum/component/ai/New(var/datum/component_container/CC)
+	..(CC)
```
Comment (PJB3005) [inline code/modules/components/ai/ai_component.dm]: This should be implicitly passed down.

File: code/modules/components/ai/ai_component.dm [hunk 3]
```diff
@@ -0,0 +1,13 @@
+/datum/component/ai
+	var/datum/component/controller/controller
+
+	var/state=0 // AI_STATE_* of the AI.
+
+/datum/component/ai/RecieveSignal(var/message_type, var/list/args)
+	switch(message_type)
+		if("state") // list("name"="statename")
```
Comment (PJB3005) [inline code/modules/components/ai/ai_component.dm]: Probably best to use defines here.

File: code/modules/components/ai/atmos.dm [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+/datum/component/ai/atmos_checker/RecieveSignal(var/message_type, var/list/args)
+	switch(message_type)
+		if("life")
+			OnLife()
+		else
+			..(message_type, args)
+
+/datum/component/ai/atmos_checker/proc/OnLife()
+	if(isliving(container.holder))
```
Comment (PJB3005) [inline code/modules/components/ai/atmos.dm]: Should do an opposite check and return here to remove indentation from the rest of the proc.

File: code/modules/components/ai/atmos.dm [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+		var/atmos_suitable = 1
+
+		var/atom/A = container.holder.loc
+
+		if(isturf(A))
+			var/turf/T = A
+			var/datum/gas_mixture/Environment = T.return_air()
+
+			if(Environment)
+				if(abs(Environment.temperature - controller.getBodyTemperature()) > 40)
```
Comment (PJB3005) [inline code/modules/components/ai/atmos.dm]: Magic number?

File: code/modules/components/ai/controller.dm [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	_state = new_state
+	SendSignal("state changed", list("state"=_state))
+
+/datum/component/controller/proc/getState()
+	return _state
+
+/datum/component/controller/proc/setBodyTemperature(var/temp)
+	SendSignal("body temp", list("temp"=temp,"from"=src))
+
+/datum/component/controller/proc/getBodyTemperature()
```
Comment (PJB3005) [inline code/modules/components/ai/controller.dm]: I don't think it makes sense to have body temperature here. There should be a separate component for it.

That does raise some issues with getting the value though, so you'll need a helper.

File: code/modules/components/ai/hostile/inject_poison.dm [hunk 7]
```diff
@@ -0,0 +1,23 @@
+/datum/component/ai/inject_poison
+	var/poison_type = "" // STOXIN, etc
+	var/poison_per_bite = 0 // Mols to inject
+	var/inject_prob = 0 // Chance to inject, -1 = ALWAYS
+	var/max_poison = 0 // Maximum mols in target's blood. 0 = INF
+
+/datum/component/ai/inject_poison/RecieveSignal(var/message_type, var/list/args)
+	switch(message_type)
+		if("attacking target") // list("target"=A)
```
Comment (PJB3005) [inline code/modules/components/ai/hostile/inject_poison.dm]: Man could you imagine if we had decorators and stuff?

File: code/modules/components/ai/mob_controller.dm [hunk 8]
```diff
@@ -0,0 +1,22 @@
+/datum/component/controller/mob
```
Comment (PJB3005) [inline code/modules/components/ai/mob_controller.dm]: That's... a really non specific component.

File: code/modules/components/component_container.dm [hunk 9]
```diff
@@ -0,0 +1,86 @@
+#define COMPONENT_ADDED "component added"
+#define COMPONENT_REMOVING "component removed"
+
+// Basic multipurpose component container.
+/datum/component_container
```
Comment (PJB3005) [inline code/modules/components/component_container.dm]: Should probably clear components on `Destroy`

File: code/modules/components/ai/door_opener.dm [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+/datum/component/ai/door_opener/proc/OnAttackingTarget(var/atom/target)
+	if(istype(target,/obj/machinery/door))
+		var/obj/machinery/door/D = target
+		if(CanOpenDoor(D))
+			if(get_dist(src, target) > 1)
+				return // keep movin'.
+			controller.setBusy(TRUE)
+			SendSignal("move", "dir"=0) // Stop movement?
+			D.visible_message("<span class='warning'>\The [D]'s motors whine as four arachnid claws begin trying to force it open!</span>")
```
Comment (ghost) [inline code/modules/components/ai/door_opener.dm]: May want to change this to something more generic than 'four arachnid claws'

**Comment by Zth--** [comment]
Will you document it?

**Comment by SangriaMonkey** [comment]
:thinking:

File: code/__HELPERS/maths.dm (hunk 1 above)
Comment (N3X15) [reply code/__HELPERS/maths.dm]: Give me a "proper" one to use that does the same thing and I'll use it.

File: code/modules/components/ai/ai_component.dm (hunk 2 above)
Comment (N3X15) [reply code/modules/components/ai/ai_component.dm]: I like passing arguments explicitly to confuse myself a bit less.  If it's a code quality requirement, I can do it.

File: code/modules/components/ai/atmos.dm (hunk 5 above)
Comment (N3X15) [reply code/modules/components/ai/atmos.dm]: Yep.  Imported from old spidercode.

File: code/modules/components/ai/mob_controller.dm (hunk 8 above)
Comment (N3X15) [reply code/modules/components/ai/mob_controller.dm]: Mob controller.  I plan on making a counterpart for obj-bots.

**Comment by N3X15** [comment]
I'll get to making these changes soonish.  Have other things on my plate, at the moment.

File: code/modules/components/ai/atmos.dm (hunk 5 above)
Comment (PJB3005) [reply code/modules/components/ai/atmos.dm]: Considering this component will be a bit more general I'd say making this a variable should be a good idea.

**Comment by FudgePucker** [comment]
So does this mean we can have better simple mob AI? Or at least that it would be easier to code simple mob AI?

**Comment by Shadowmech88** [comment]
It'll be easier to tailor mob AI because the AI will be formed out of a combination of components that each have a specific function, rather than having to copy+paste tons of code from a bunch of different mobs into your mob's `Life()` to get it to behave the way you want.

File: code/modules/components/ai/hostile/inject_poison.dm (hunk 7 above)
Comment (N3X15) [reply code/modules/components/ai/hostile/inject_poison.dm]: Could probably do it with a precompile step, but that starts getting into "clusterfuck of a buildsystem" territory.

**Comment by N3X15** [comment]
Started documentation here.  http://ss13.moe/wiki/index.php/User:N3X15/Components

**Comment by Probe1** [comment]
you go girl

**Comment by N3X15** [comment]
Okay, the basic shit is working.  We have a spider that attacks, injects, and has working target priority (meaning it already works better than old spiders).  I'd say the framework is done.

**Comment by Probe1** [comment]
i'll look over it and merge in 15 hours.

**Comment by ghost** [comment]
>15 hours ago

@Probe1 the time is now

**Comment by optimumtact** [comment]
garbage

**Comment by Probe1** [comment]
why tho

**Pull request merged**